docs/guide/api-environment-frameworks.md

```diff
@@ -0,0 +1,410 @@
+
+:::info transformRequest naming
+We are using `transformRequest(url)` and `warmupRequest(url)` in the current version of this proposal so it is easier to discuss and understand for users used to Vite's current API. Before releasing, we can take the opportunity to review these names too. For example, it could be named `environment.processModule(url)` or `environment.loadModule(url)` taking a page from Rollup's `context.load(id)` in plugin hooks. For the moment, we think keeping the current names and delaying this discussion is better.
+:::
+
+Vite also exposes a `RunnableDevEnvironment`. While this is only implementable for some runtimes as it requires the runtime to be the same with the one the Vite server is running in, this works similarly with `ssrLoadModule`. You can guard any runnable environment with an `isRunnableDevEnvironment` function.
+In dev mode the default `ssr` environment is a `RunnableDevEnvironment`.
+
+```ts
+export class RunnableDevEnvironment extends DevEnvironment {
+  public readonly runner: ModuleRunnner
+}
+
+class ModuleRunner {
+  /**
+   * URL to execute. Accepts file path, server path, or id relative to the root.
+   * Returns an instantiated module (same as in ssrLoadModule)
+   */
+  public async import(url: string): Promise<Record<string, any>>
+  /**
+   * Other ModuleRunner methods...
+   */
+}
+
+if (isRunnableDevEnvironment(server.environments.ssr)) {
+  await server.environments.ssr.runner.import('/entry-point.js')
+}
+```
+
+:::info
+In the v5.1 Runtime API, there were `executeUrl` and `executeEntryPoint` methods - they are now merged into a single `import` method. If you want to opt-out of the HMR support, create a runner with `hmr: false` flag.
+:::
+
+## Using custom environments
+
+A Vite dev server exposes two environments by default: a `client` environment and an `ssr` environment. The client environment is a browser environment by default and the SSR environment runs in the same Node runtime as the Vite server by default.
+
+To register a custom environment (for example to have a separate module graph for [RSC](https://react.dev/blog/2023/03/22/react-labs-what-we-have-been-working-on-march-2023#react-server-components)), you can pass a new property to the `environments` option. You can also use a plugin (See [Environment API for Plugins](./api-environment-plugins.md#registering-new-environments-using-hooks)).
+
+```ts
+import { createServer, createRunnableDevEnvironment } from 'vite'
+
+const server = await createServer({
+  /* ... */
+  environments: {
+    rsc: {
+      dev: {
+        createEnvironment(name, config) {
+          return createRunnableDevEnvironment(name, config)
+        },
+      },
+      build: {
+        outDir: '/dist/rsc',
+      },
+    },
+  },
+})
+
+// TODO: Maybe we should expose a `createNodeEnvironment` function (an environment factory for Node)?
```

Comment:
Ideally, I think framework authors should be able to just use environment factories every where rather than using `createEnvironment` function.
For example:
```ts
const server = await createServer({
  /* ... */
  environments: {
    rsc: createNodeEnvironment({
      build: {
        outDir: '/dist/rsc',
      },
    }),
    rscEdge: createWorkerdEnvironment({
      build: {
        outDir: '/dist/rsc-edge',
      },
    }),
  },
})
```

Reply:
I was reading through the docs and have the exact same thought too 😄 This way frameworks and users should never need to know about `createEnvironment`. I wonder if it's also worth never exposing this type to users, e.g.:

```ts
createNodeEnvironment(): ExtendedEnvironmentOptions // has `createEnvironment`

interface UserConfig {
  environments: {
    [key: string]: EnvironmentOptions
  }
}

```

Reply:
Ya, the idea has been for a while that environment providers should be the only ones that need to use `createEnvironment` (and maybe tools like Vitest). Having all the docs in a single page was not making this clear at all.

How would that types work @bluwy? Do we cast it internally? I think it is enough to document that this is a very low level tool that is only intended for environment providers.

Reply:
Yeah I'd assume we would cast it internally. We don't have to do it now, but I feel like it can be one of the ways to minimize the API surface area a bit.